streams/shared/src/main/scala/scalaz/stream/ZStream.scala

```diff
@@ -116,74 +178,102 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
+            innerFailure <- Promise.make[E1, Nothing].toManaged_
+            innerFibers  <- Ref.make[List[Fiber[E1, Unit]]](Nil).toManaged_
+            // - The driver stream forks an inner fiber for each stream created
+            //   by f, with an upper bound of n concurrent fibers, enforced by the semaphore.
+            //   - On completion, the driver stream joins all pending inner fibers:
+            //     - If one of them failed, all other fibers are interrupted
+            //     - If they all succeeded, Take.End is enqueued
+            //   - On error, the driver stream interrupts all inner fibers and emits a
+            //     Take.Fail value
+            //   - On interruption, the driver stream interrupts all inner fibers and
+            //     shuts down the queue
+            // - Inner fibers enqueue Take values from their streams to the output queue
+            //   - On error, an inner fiber enqueues a Take.Error value and signals its failure
+            //     with a promise. The driver will pick that up and interrupt all other fibers.
+            //   - On interruption, an inner fiber does nothing
+            //   - On completion, an inner fiber does nothing
+            driver = ZManaged
+              .unwrap(self.fold[R2, E1, A, List[Fiber[E1, Unit]]].map { fold =>
+                fold(
+                  Nil,
+                  _ => true,
+                  (fibers, a) =>
+                    permits.withPermit {
+                      f(a)
+                        .foreach(b => out.offer(Take.Value(b)).unit)
+                        .catchAll(e => out.offer(Take.Fail(e)) *> innerFailure.fail(e) *> ZIO.fail(e))
+                        .fork
+                        .map(_ :: fibers)
+                    }
+                )
+              })
+              .mapM(fibers => innerFibers.set(fibers).const(fibers))
+              .foldCauseM(
+                _.failureOrCause.fold(
+                  // The driver stream failed.
+                  e => out.offer(Take.Fail(e)).unit.toManaged_,
+                  // The driver stream had a defect.
+                  c => (out.shutdown *> ZIO.halt(c)).toManaged_
+                ),
+                innerFibers =>
+                  innerFailure.await
+                    .raceWith(Fiber.joinAll(innerFibers))(
+                      // One of the inner fibers failed. It already enqueued its failure,
+                      // so we don't need to do anything except interrupt all the other fibers.
+                      leftDone = (_, _) => Fiber.interruptAll(innerFibers),
+                      // All fibers completed successfully, so we just need to signal that
+                      // we're done.
+                      rightDone = (_, failureAwait) => out.offer(Take.End) *> failureAwait.interrupt
+                    )
+                    .toManaged_
+              )
+              .fork
+            supervisingDriver = ZManaged {
```

Comment:
@jdegoes This step is pretty clunky; we need to prepend a step to the finalizer of `driver` here, which makes me feel like we're missing an operator for that on `ZManaged`.

Reply:
Yes, I think so. We're missing `finalizer`, something like:

```scala
def finalizer(f: UIO[_]): ZManaged[Any, Nothing, Unit]
```

Then we can do something like `zmanaged.flatMap(a => ZManaged.finalizer(fin).const(a))`

That would do it, right?

Reply:
Yes! Thank you. That's what I was looking for.